frappe/__init__.py

```diff
@@ -204,7 +204,8 @@ def connect_replica():
 	local.replica_db = get_db(host=local.conf.replica_host, user=user, password=password, port=port)
 
 	# swap db connections
-	local.primary_db = local.db
+	local.db.close()
```

Comment:
This is already done on line 550

Reply:
@akhilnarang I thought the one on line 550 closes the `replica_db` connection in `local.db` before switching back to the `primary_db` while this update closes the `primary_db` connection in `local.db` before switching to the `replica_db`

Reply:
The code you linked in the issue is from the `develop` branch, you've made this PR to `version-12-hotfix` 

When you call a function with read_only decorator and have a replica set up, it will
- Open a connection to the replica DB stored in `local.replica_db`
- Store the current connection `local.db` into `local.primary_db`
- Store the `local.replica_db` connection into `local.db`
- <perform operations>
- Close connection in `local.db` (replica) if `local.primary_db` exists, restore primary connection to `local.db`

Reply:
Can you please update your issue with more details?
Like what error you're facing? Also if you're on version 12 - ideally please update to a supported version.

Reply:
> The code you linked in the issue is from the `develop` branch, you've made this PR to `version-12-hotfix`
> 
> When you call a function with read_only decorator and have a replica set up, it will
> 
> * Open a connection to the replica DB stored in `local.replica_db`
> * Store the current connection `local.db` into `local.primary_db`
> * Store the `local.replica_db` connection into `local.db`
> * perform operations
> * Close connection in `local.db` (replica) if `local.primary_db` exists, restore primary connection to `local.db`

Yes true but in Python if I say for example `x = 9` and then i assign a variable `a`  to  `x` (i.e `a = x`). At this point, `a` is 9. Then if I assign a variable `x` to 3 (i.e `x = 3`) that means x would be 3 but also would be 3 and not 9.

```
>>> x = 9
>>> a = x
>>> a # returns 9
>>> x = 3
>>> x # returns 3
>>> a # returns 3 because it's pointer has changed
```

In this scenario, `x` is `local.db`, `a` is `local.primary_db` and `x = 3` is similar to `local.db = local.replica_db`

Reply:
@akhilnarang 
If you want to recreate the issue I'm trying to fix do the following:

Set the `replica_host` to a replica database and `read_from_replica` as 1, 
Then try to send a report with `Auto Email Report` since to temporarily switches to the `read_only` database when pulling the data.

Reply:
![Uploading image (6).png…]()